File: core/src/com/desitum/library/game/AssetManager.java
package com.desitum.library.game;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.NinePatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.desitum.library.drawing.Drawable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Created by kodyvanry on 5/11/17.
 */

public class AssetManager {

    private static AssetManager instance;

    private List<Texture> mTextureList;
    private HashMap<Integer, Drawable> mDrawableMap;

    private AssetManager() {
        mTextureList = new ArrayList<Texture>();
        mDrawableMap = new HashMap<Integer, Drawable>();
//        new Texture(new FileHandle(Gdx.files.internal()))
    }

    public static AssetManager getInstance() {
        if (instance == null)
            instance = new AssetManager();
        return instance;
    }

    public void addTexture(Texture texture) {
        mTextureList.add(texture);
    }

    public void addTexture(String texture) {
        this.addTexture(new Texture(texture));
    }

    public Texture getTexture(int text) {
        return mTextureList.get(text);
    }

    public void addDrawable(int key, TextureRegion textureRegion) {
        mDrawableMap.put(key, new Drawable(textureRegion));
    }

    public void addDrawable(int key, Texture texture) {
        mDrawableMap.put(key, new Drawable(texture));
    }

    public void addDrawable(int key, NinePatch ninePatch) {
        mDrawableMap.put(key, new Drawable(ninePatch));
    }

    public void addDrawable(int key, Drawable drawable) {
        mDrawableMap.put(key, drawable);
    }

    public Drawable getDrawable(int key) {
        return mDrawableMap.get(key);
    }

    public static void dispose() {
        AssetManager assetManager = getInstance();
        for (Texture texture : assetManager.mTextureList) {
            texture.dispose();
        }
        assetManager.mTextureList = null;
        assetManager = null;
        instance = null;
    }
}

File: core/src/com/desitum/library/game/AssetManager.kt
package com.desitum.library.game

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.NinePatch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.desitum.library.drawing.Drawable
import java.util.*

/**
 * Created by kodyvanry on 5/11/17.
 */

class AssetManager private constructor() {

    private var mTextureList: MutableList<Texture> = ArrayList<Texture>()
    private val mDrawableMap: HashMap<Int, Drawable> = HashMap<Int, Drawable>()
    private val mBitmapFonts: HashMap<Int, BitmapFont> = HashMap<Int, BitmapFont>()

    fun addTexture(texture: Texture) {
        mTextureList.add(texture)
    }

    fun addFont(key: Int, bitmapFont: BitmapFont) {
        mBitmapFonts.put(key, bitmapFont)
    }

    fun addTexture(texture: String) {
        this.addTexture(Texture(texture))
    }

    fun getTexture(text: Int): Texture {
        return mTextureList[text]
    }

    fun getFont(key: Int): BitmapFont {
        return mBitmapFonts[key]!!
    }

    fun addDrawable(key: Int, textureRegion: TextureRegion) {
        mDrawableMap.put(key, Drawable(textureRegion))
    }

    fun addDrawable(key: Int, texture: Texture) {
        mDrawableMap.put(key, Drawable(texture))
    }

    fun addDrawable(key: Int, ninePatch: NinePatch) {
        mDrawableMap.put(key, Drawable(ninePatch))
    }

    fun addDrawable(key: Int, drawable: Drawable) {
        mDrawableMap.put(key, drawable)
    }

    fun getDrawable(key: Int): Drawable {
        return Drawable(mDrawableMap[key].let { it!! })
    }

    companion object {

        @JvmStatic val instance: AssetManager by lazy { AssetManager() }

        @JvmStatic fun dispose() {
            val assetManager: AssetManager = instance
            for (texture in assetManager.mTextureList) {
                texture.dispose()
            }
        }
    }
}
